File: frontend/src/api/rooms.ts
import {
  EditRoomFormValues,
  RoomFormValues,
} from "@/features/rooms/components/room-form";
import { api } from "./api";

export const getRooms = async () => {
  const response = await api.get("/rooms");
  if (response.status != 200) {
    throw new Error("Failed to fetch rooms");
  }
  return response.data as Room[];
};

export const createRoom = async (values: RoomFormValues, images: string[]) => {
  const response = await api.post("/rooms", { ...values, images });
  if (response.status != 201) {
    throw new Error("Failed to create room");
  }
  return response.data;
};

export const updateRoom = async (
  id: string,
  values: EditRoomFormValues,
  images: string[]
) => {
  const response = await api.patch(`/rooms/${id}`, { ...values, images });
  if (response.status != 200) {
    throw new Error("Failed to update room");
  }
  return response.data;
};

export const deleteRoom = async (id: string) => {
  const response = await api.delete(`/rooms/${id}`);
  if (response.status != 200) {
    throw new Error("Failed to delete room");
  }
  return response.data;
};

export const getBookingRooms = async (room_type: string) => {
  const response = await api.get(`/rooms/book?room_type=${room_type}`);
  if (response.status != 200) {
    throw new Error("Failed to fetch rooms");
  }
  return response.data as Room[];
};

File: internal/database/rooms.go
package database

import (
	"hms-api/models"
	"hms-api/types"

	"github.com/google/uuid"
)

func (s *service) GetRooms() ([]models.Room, error) {
	var rooms []models.Room
	result := s.db.Model(&models.Room{}).Find(&rooms)
	return rooms, result.Error
}

func (s *service) CreateRoom(payload *types.CreateRoomPayload) error {
	newRoom := &models.Room{
		RoomType:    (*models.RoomType)(&payload.RoomType),
		RoomNumber:  payload.RoomNumber,
		Price:       payload.Price,
		BedType:     (*models.BedType)(&payload.BedType),
		MaxCapacity: payload.MaxCapacity,
		Amenities:   payload.Amenities,
		Images:      payload.Images,
	}
	result := s.db.Model(&models.Room{}).Create(newRoom)
	return result.Error
}

func (s *service) GetRoom(id uuid.UUID) (*models.Room, error) {
	var room models.Room
	result := s.db.Model(&models.Room{}).Where("id = ?", id).First(&room)
	return &room, result.Error
}

func (s *service) UpdateRoom(id uuid.UUID, payload *types.UpdateRoomPayload) error {
	_updates := types.ParseRoomUpdates(payload)
	result := s.db.Model(&models.Room{}).Where("id = ?", id).Updates(_updates)
	return result.Error
}

func (s *service) DeleteRoom(id uuid.UUID) error {
	result := s.db.Delete(&models.Room{}, id)
	return result.Error
}
